### PR TITLE
Full parameter options and licensing

### PR DESCRIPTION
**Pull request by aelanman**
### Summary:
- Includes the full options to specify frequency and time arrays as used in pyuvsim.
- Adds BSD 3-clause license and a changelog in preparation for first release.
- Style change --- Ordering import statements
- Some changes that enable simulations to run with the minimum required parameters.
- New scripts

### Frequency/time param explanation.
How best to define the time and frequency arrays is a matter of some debate among the datasim team. For instance -- if each frequency channel has a center frequency and a channel width, then the full bandwidth (that is, the amount of frequency space covered by the channels) should be `(Nfreqs +1) * channel_width`, where the +1 accounts for the extra half channels above the last and below the first frequency centers.

The debate is whether this is sensible in a simulation context. Certainly if we want to compare with data, the frequency channel centers should match the channel centers of a data file. But how do we define the bandwidth in either case? Is bandwidth even meaningful for a simulator?

Overall -- This PR sets the following convention, using the parse_..._params() functions borrowed from pyuvsim:
~~Bandwidth = channel_width  * (Nfreqs + 1)~~
~~freq_array = (f_0, f_1, ..., f_Nfreqs) where f_i = f_0 + i * channel_width~~
and similarly for the time array.

I've included functions in `utils` that will tell you what the parameters are for a given frequency or time array, and a unit test that checks that those parameters return the same array when run through the parser.

### Skymodel frequencies
I also made the change that the simulator will error if a loaded sky model doesn't have frequencies that match the obsparam frequencies. I think this is the preferable behavior, since the user may not notice that the simulation is going forward with settings different than what was specified. This can probably be downgraded to a warning instead, but it shouldn't happen silently.

### New scripts
Still a little undercooked... these will be better made in a future PR. I've added a script that generates a random antenna layout of a given size and maximum baseline, and another that makes a HEALPix map of evenly spaced points whose brightnesses correspond with their declinations. I want to make an image of a recognizable pattern on the sky and confirm that sky rotation and positioning look reasonable.

Sorry for the long explanation, and I hope this doesn't disrupt anything you're working on!

__[Edit 3/29]:__
Frequencies are considered channel centers.
Even though bandwidth and channel_width are non-meaningful parameters for a simulation that __samples__ the sky rather than integrates, we use these terms consistent with how observers use them. Therefore, `bandwidth = Nfreqs * channel_width` and `freq_array = np.linspace(start_freq, start_freq + bandwidth, Nfreqs, endpoint=False)`.

### CODE REVIEW

File: healvis/beam_model.py [hunk 1]
```diff
@@ -316,6 +322,8 @@ def beam_val(self, az, za, freqs, **kwargs):
         """
         if self.beam_type == 'uniform':
             if isinstance(az, np.ndarray):
+                if np.isscalar(freqs):
```
Comment (nkern) [inline healvis/beam_model.py]: I don't think this is doing what you expect b/c its nested within an effective `is np.ndarray`

File: healvis/data/configs/obsparam_test.yaml [hunk 2]
```diff
@@ -13,9 +13,8 @@ telescope:
   telescope_location: (-30.72152777777791, 21.428305555555557, 1073.0000000093132)  # LatLonAlt location
   telescope_name: HERA
 freq:
-  Nfreqs: 3
-  start_freq: 100000000.0  # Hz
-  bandwidth:  100000000.0  # Hz
+  freq_array: [1.00e+08, 1.05e+08, 1.10e+08, 1.15e+08, 1.20e+08,
```
Comment (nkern) [inline healvis/data/configs/obsparam_test.yaml]: I had it in `run_simulation_partial_freq` that you can feed either 
1) `Nfreqs`, `start_freq` and `bandwidth` or
2) `freq_array` and if the latter was fed it superceded the former. can we keep this here?

**Comment by nkern** [comment]
@aelanman I disagree that `bandwidth = chan_width * (Nfreqs + 1)`. Furthermore, the frequencies that come from the correlator and that of a UVData object are, to my knowledge, the channel start. Therefore, bandwidth = `Nfreqs` * `channel_width` and `freq_array` should be defined as
`np.linspace(start_freq, start_freq + bandwidth, Nfreqs, endpoint=False)`

It seems like we are perhaps thinking too hard about this? It seems like the question is that a frequency is defined by its bin center or bin start, in both cases `freq_array` and `bandwidth` are defined as I have above, its just that in the latter case we would give `start_time += chan_width / 2` relative to the former.

Long story short, I don't think we should change our definition of bandwidth or Nfreqs in order to conform to a bin start or bin center convention, which should only change the values of `freq_array`, not their total extent or number of elements. Let me know what your thoughts are on this.

File: healvis/beam_model.py (hunk 1 above)
Comment (aelanman) [reply healvis/beam_model.py]: This came up when I was running a single-frequency simulation with a uniform beam. There's no check that `freqs` is an ndarray, even if az/za are.

File: healvis/data/configs/obsparam_test.yaml (hunk 2 above)
Comment (nkern) [reply healvis/data/configs/obsparam_test.yaml]: and of course we should make that clear via comments

File: healvis/data/configs/obsparam_test.yaml [hunk 3]
```diff
@@ -13,9 +13,8 @@ telescope:
...
+                1.25e+08, 1.30e+08, 1.35e+08, 1.40e+08, 1.45e+08]
 time:
   Ntimes: 5
   time_cadence: 100.0  # seconds
```
Comment (nkern) [inline healvis/data/configs/obsparam_test.yaml]: also idk if you noticed this, but I switched our vocabular from `integration_time` to `time_cadence`. This is perhaps what you were referring to in "what does a bandwidth [or integration similarly] mean for a simulator?" in that we aren't _integrating_ data as the correlator is doing, we are _sampling_. For that reason I didn't like the usage of `integration_time` to set the time_array spacing, so I opted to use `time_cadence` instead, and now the `integration_time` array in the output `UVData` is just set to one by default.

File: healvis/simulator.py [hunk 4]
```diff
@@ -84,56 +93,164 @@ def parse_telescope_params(tele_params):
...
 
     Returns:
-        dictionary
-            | Nfreqs : int
-            | channel_width : float, [Hz]
-            | freq_array : 2D ndarray, shape (1, Nfreqs) [Hz]
+        dict of array properties:
+            |  channel_width: (float) Frequency channel spacing in Hz
+            |  Nfreqs: (int) Number of frequencies
+            |  freq_array: (dtype float, ndarray, shap=(Nspws, Nfreqs)) Frequency channel centers in Hz
```
Comment (nkern) [inline healvis/simulator.py]: *shape

File: healvis/simulator.py [hunk 5]
```diff
@@ -84,56 +93,164 @@ def parse_telescope_params(tele_params):
...
+                           time_params['end_time'] + inttime_days,
+                           time_params['Ntimes'], endpoint=False)
+
+    if time_params['Ntimes'] != 1:
+        assert np.allclose(np.diff(time_arr), inttime_days * np.ones(time_params["Ntimes"] - 1), atol=dayspersec)   # To nearest second
+
+    return_dict['time_cadence'] = time_params['time_cadence']
+    return_dict['time_array'] = time_arr
+    return_dict['Ntimes'] = time_params['Ntimes']
+    return_dict['Nspws'] = 1
```
Comment (nkern) [inline healvis/simulator.py]: I see this is copy/paste from the `pyuvsim` `simsetup` module. I left these lines out on purpose b/c I think they do more harm then good. I don't see a point in returning a notion of `spws` and `pols` when they weren't requested for and aren't needed in the context of how this function is used. I would opt to say we don't return populate these keys.

File: healvis/simulator.py [hunk 6]
```diff
@@ -148,13 +265,14 @@ def complete_uvdata(uv_obj, run_check=True):
...
     time_array = uv_obj.time_array      # (Ntimes,)
+    dt = np.diff(np.unique(time_array))[0]
 
     uv_obj.baseline_array = np.tile(bl_array, uv_obj.Ntimes)
     uv_obj.ant_1_array, uv_obj.ant_2_array = uv_obj.baseline_to_antnums(uv_obj.baseline_array)
     uv_obj.Nbls = np.unique(uv_obj.baseline_array).size
     uv_obj.Nblts = uv_obj.Nbls * uv_obj.Ntimes
     uv_obj.time_array = np.repeat(time_array, uv_obj.Nbls)
-    uv_obj.integration_time = np.repeat(np.ones_like(time_array), uv_obj.Nbls)
+    uv_obj.integration_time = np.repeat(np.ones_like(time_array), uv_obj.Nbls) * dt * 24 * 3600.    # Seconds
```
Comment (nkern) [inline healvis/simulator.py]: As I said above, I think in the context of a simulator it isn't fair to set the integration time to be simply the `dt` of the `time_array`. TBH it should probably be zero or `NaN` but I think that would cause us more headache as downstream modules might break. I think setting it to one is a fair default for the time being..

File: healvis/simulator.py [hunk 7]
```diff
@@ -385,15 +506,18 @@ def run_simulation(param_file, Nprocs=1, sjob_id=None, add_to_history=''):
         param_dict = param_file
```
Comment (nkern) [inline healvis/simulator.py]: we should do a deepcopy here to prevent accidental inplace operations further down.

File: healvis/simulator.py [hunk 8]
```diff
@@ -403,15 +527,25 @@ def run_simulation(param_file, Nprocs=1, sjob_id=None, add_to_history=''):
...
+    savepath = None
+    if 'savepath' in skyparam:
+        savepath = skyparam.pop('savepath')
+
+    sky = sky_model.construct_skymodel(sky_type, **skyparam)
+
+    # If loading a healpix map from disk, confirm its frequencies match the obsparam frequencies.
+    if sky_type.lower() not in ['flat_spec', 'gsm']:
+        try:
+            assert np.all(freq_array == sky.freqs)
```
Comment (nkern) [inline healvis/simulator.py]: I don't like `==` for float data types. I agree that if they were literally the same bytes then this will work, but I think that if they disagreed at 1e-15 level then this should pass, and in the way its written it will fail. I think `np.isclose(arr1, arr2).all()` would be better, personally.

File: healvis/simulator.py [hunk 9]
```diff
@@ -436,11 +573,13 @@ def run_simulation(param_file, Nprocs=1, sjob_id=None, add_to_history=''):
...
-    pols = beam_attr.pop("pols")
-    beam_freq_interp = beam_attr.pop("beam_freq_interp")
-    smooth_beam = beam_attr.pop("smooth_beam")
-    smooth_scale = beam_attr.pop("smooth_scale")
-    obs = setup_observatory_from_uvdata(uv_obj, fov=param_dict['beam'].pop("fov"), set_pointings=True,
+
+    pols = beam_attr.pop("pols", None)
+    beam_freq_interp = beam_attr.pop("beam_freq_interp", 'cubic')
+    smooth_beam = beam_attr.pop("smooth_beam", False)
+    smooth_scale = beam_attr.pop("smooth_scale", 2.0)
```
Comment (nkern) [inline healvis/simulator.py]: Even though I agree that `2.0` is the default in the function, I think this should be `smooth_scale = beam_attr.pop("smooth_scale", 2.0)`, because if someone isn't smoothing it won't be used anyways, but if they _are_ smoothing and forgot to specify the scale, then this will raise an error downstream, as one would hope if they user forgot to specify a crucial parameter.

File: healvis/beam_model.py (hunk 1 above)
Comment (nkern) [reply healvis/beam_model.py]: oh sorry, i misread the line above as `freqs` too.

File: healvis/simulator.py (hunk 7 above)
Comment (aelanman) [reply healvis/simulator.py]: I don't think we actually overwrite any entries in the `param_dict`, but it doesn't hurt to be extra careful.

File: healvis/simulator.py (hunk 6 above)
Comment (aelanman) [reply healvis/simulator.py]: That's a very good point. In pyuvsim, we've been using integration_time as a proxy for timestep/time_cadence, but it's really the wrong term for it.

I'm going to set it to NaN, though. Setting it to 1 works, but if anything downstream uses the integration time it will introduce a silent error.

File: healvis/data/configs/obsparam_test.yaml (hunk 2 above)
Comment (aelanman) [reply healvis/data/configs/obsparam_test.yaml]: I've put it back as `Nfreqs`, `start_freq`, and `bandwidth`. It's still consistent with the freqs in the gsm_nside32 test file.

File: healvis/data/configs/obsparam_test.yaml (hunk 3 above)
Comment (aelanman) [reply healvis/data/configs/obsparam_test.yaml]: I agree, it's a much better term for what we're doing. The change in terminology should be finished throughout.

File: healvis/simulator.py (hunk 6 above)
Comment (aelanman) [reply healvis/simulator.py]: @nkern  It seems that using NaNs causes some deep issues with the partial_freq methods (since nan != nan by python logic). We're going to discuss these parameter definitions at the datasim telecon tomorrow, if you'd like to join. For now we'll go with your idea of using 0's. At least that accurately describes the total time being integrated!

Comment (nkern) [reply healvis/simulator.py]: Hmm, I guess I'm fine with that for now in this PR, but I do think that might cause some trouble later on, as I believe we have some analysis modules downstream that flags data with integration time = 0, but I guess we'll cross the bridge when we get there.

File: healvis/data/configs/obsparam_test.yaml [hunk 10]
```diff
@@ -13,9 +13,10 @@ telescope:
   telescope_location: (-30.72152777777791, 21.428305555555557, 1073.0000000093132)  # LatLonAlt location
   telescope_name: HERA
 freq:
-  Nfreqs: 3
-  start_freq: 100000000.0  # Hz
-  bandwidth:  100000000.0  # Hz
+  bandwidth: 50000000.0    # Total frequency coverage (Nfreqs * channel_width)
+  start_freq: 100000000.0   # Center of first frequency channel
+  channel_width: 5000000.0  # Separation of frequency channel centers
+  Nfreqs: 10                # Number of frequency channels
```
Comment (nkern) [inline healvis/data/configs/obsparam_test.yaml]: I think we should specify three of the following, not four, because they can produce conflicting information if not properly set.
I think the most natural thing is `start_freq`, `bandwidth` and `Nfreqs`, as we had before. so prob we should get rid of `channel_width`. alternatively, if you prefer `channel_width`, then get rid of `bandwidth`.

File: healvis/simulator.py [hunk 11]
```diff
@@ -84,56 +93,162 @@ def parse_telescope_params(tele_params):
...
+                     'channel_width', 'bandwidth']
+    fa, sf, ef, nf, cw, bw = [fk in freq_params for fk in freq_keywords]
+    kws_used = ", ".join(freq_params.keys())
+
+    if fa:
+        freq_arr = np.array(freq_params['freq_array'])
+        freq_params['Nfreqs'] = freq_arr.size
+        if freq_params['Nfreqs'] > 1:
+            freq_params['channel_width'] = np.diff(freq_arr)[0]
+            assert np.all(np.diff(freq_arr) == freq_params['channel_width'])
```
Comment (nkern) [inline healvis/simulator.py]: should be `isclose` instead of `==` here too I think

**Comment by nkern** [comment]
@aelanman I've got some suggestions for the `parse_` functions in `simulator.py`. In general, I don't like the inplace editing of the dictionaries which will cause trouble when passing dictionaries whose values will be overwritten silently. I also think its easier to enforce a policy of parsing order on the multiple keys that can be specified, rather than enforce that all keys are compliant with each other.
Also added extensive testing for the parsing functions.

If you disagree with this change we can revert the commit: Let me know what your thoughts are on this in the meantime.

**Comment by aelanman** [comment]
@nkern I think the change makes sense and makes it a little clearer what's happening. I've put back in the assertion tests on the final freq/time arrays, because I don't think it hurts to be extra sure that the array and parameters are consistent.

I've also added a test on the redundancy selection. The actual redundancy selection methods used by pyuvdata are well-tested there, but this is to make sure the selection is understood properly.

**Comment by nkern** [comment]
@aelanman looks like an error in your redundancy test. once that is fixed should be good to go. Also, do you want to make a commit that bumps the `VERSION` file to `1.0.0`, that way once we merge this in we can tag a a `v1.0.0` release?
